src/uct/ib/dc/base/dc_iface.c

```diff
@@ -307,3 +307,180 @@ ucs_status_t uct_dc_iface_flush(uct_iface_h tl_iface, unsigned flags, uct_comple
+    }
+    /* We do not have any peer address at this point, so init basic subclasses
+     * only (for statistics, iface, etc) */
+    status = UCS_CLASS_INIT(uct_base_ep_t, (void*)(&ep->super),
+                            &iface->super.super.super);
+    if (status != UCS_OK) {
+        ucs_error("Failed to initialize fake FC ep, status: %s",
+                  ucs_status_string(status));
+        goto err_free;
+    }
+
+    status = uct_dc_ep_basic_init(iface, ep);
+    if (status != UCS_OK) {
+        ucs_error("FC ep init failed %s", ucs_status_string(status));
+        goto err_cleanup;
+    }
+
+    iface->tx.fake_fc_ep = ep;
+    return UCS_OK;
+
+err_cleanup:
+    UCS_CLASS_CLEANUP(uct_base_ep_t, &ep->super);
+err_free:
+    ucs_free(ep);
+err:
+    return status;
+}
+
+void uct_dc_iface_cleanup_fc_ep(uct_dc_iface_t *iface)
+{
+    uct_dc_ep_pending_purge(&iface->tx.fake_fc_ep->super.super, NULL, NULL);
+    ucs_arbiter_group_cleanup(&iface->tx.fake_fc_ep->arb_group);
+    uct_rc_fc_cleanup(&iface->tx.fake_fc_ep->fc);
+    UCS_CLASS_CLEANUP(uct_base_ep_t, iface->tx.fake_fc_ep);
+    ucs_free(iface->tx.fake_fc_ep);
+}
+
+void uct_dc_iface_cleanup_pending_req(uct_pending_req_t *req)
+{
+    uct_rc_fc_request_t *freq = ucs_container_of(req, uct_rc_fc_request_t,
+                                                 req);
+    uct_dc_fc_request_t *dreq = ucs_derived_of(freq, uct_dc_fc_request_t);
+    ibv_destroy_ah(dreq->ibv_ah);
+    ucs_mpool_put(freq);
+}
+
+ucs_status_t uct_dc_iface_create_ah(uct_dc_iface_t *dc_iface, uint16_t lid,
+                                    struct ibv_ah **ah_p)
+{
+    struct ibv_ah_attr ah_attr;
+    struct ibv_ah *ah;
+    uct_ib_iface_t *iface = &dc_iface->super.super;
+
+    /* TODO: GRH, path_bits, etc */
+    ah_attr.sl            = iface->config.sl;
+    ah_attr.src_path_bits = iface->path_bits[0];
+    ah_attr.dlid          = lid | ah_attr.src_path_bits;
+    ah_attr.port_num      = iface->config.port_num;
+    ah_attr.is_global     = 0;
+    ah_attr.static_rate   = 0;
+
+    ah = ibv_create_ah(uct_ib_iface_md(iface)->pd, &ah_attr);
+    if (ucs_unlikely(ah == NULL)) {
+        ucs_error("Failed to create ah on "UCT_IB_IFACE_FMT,
+                  UCT_IB_IFACE_ARG(iface));
+        return UCS_ERR_INVALID_ADDR;
+    }
+
+    *ah_p = ah;
+    return UCS_OK;
+}
+
+ucs_status_t uct_dc_iface_fc_grant(uct_pending_req_t *self)
+{
+    uct_rc_fc_request_t *freq = ucs_container_of(self, uct_rc_fc_request_t, req);
+    uct_dc_ep_t *ep           = ucs_derived_of(freq->ep, uct_dc_ep_t);
+    uct_rc_iface_t *iface     = ucs_derived_of(ep->super.super.iface,
+                                               uct_rc_iface_t);
+    uct_rc_iface_ops_t  *ops  = ucs_derived_of(iface->super.ops,
+                                               uct_rc_iface_ops_t);
+    ucs_status_t status;
+
+    ucs_assert(iface->config.fc_enabled);
+
+    status = ops->fc_ctrl(&ep->super.super, UCT_RC_EP_FC_PURE_GRANT, freq);
```

Comment:
i would add uct_rc_iface_fc_ctrl(iface, ..) helper function to remove some of the code duplication